system/default_ad_api/src/vehicle.cpp

```diff
@@ -0,0 +1,178 @@
+{
+  vehicle_status_.steering_tire_angle = msg_ptr->steering_tire_angle;
+}
+
+void VehicleNode::gear_status(const GearReport::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.gear.status = mapping(gear_type_, msg_ptr->report, ApiGear::UNKNOWN);
+}
+
+void VehicleNode::turn_indicator_status(const TurnIndicatorsReport::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.turn_indicators.status =
+    mapping(turn_indicator_type_, msg_ptr->report, ApiTurnIndicator::UNKNOWN);
+}
+
+void VehicleNode::hazard_light_status(const HazardLightsReport::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.hazard_lights.status =
+    mapping(hazard_light_type_, msg_ptr->report, ApiHazardLight::UNKNOWN);
+}
+
+void VehicleNode::map_projector_info(const MapProjectorInfo::ConstSharedPtr msg_ptr)
+{
+  map_projector_info_ = msg_ptr;
+}
+
+void VehicleNode::energy_status(
+  const vehicle_interface::EnergyStatus::Message::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.energy_percentage = msg_ptr->energy_level;
+}
+
+void VehicleNode::on_timer()
+{
+  vehicle_status_.stamp = now();
+  autoware_ad_api::vehicle::VehicleKinematics::Message vehicle_kinematics;
+  vehicle_kinematics.pose.header = kinematic_state_msgs_->header;
+  vehicle_kinematics.pose.pose = kinematic_state_msgs_->pose;
+  vehicle_kinematics.twist.header = kinematic_state_msgs_->header;
+  vehicle_kinematics.twist.header.frame_id = kinematic_state_msgs_->child_frame_id;
+  vehicle_kinematics.twist.twist = kinematic_state_msgs_->twist;
+  if (map_projector_info_->type == "MGRS") {
+    lanelet::GPSPoint projected_gps_point = lanelet::projection::MGRSProjector::reverse(
+      toBasicPoint3dPt(kinematic_state_msgs_->pose.pose.position), map_projector_info_->mgrs_grid);
+    vehicle_kinematics.geographic_pose.header = kinematic_state_msgs_->header;
+    vehicle_kinematics.geographic_pose.header.frame_id = "global";
+    vehicle_kinematics.geographic_pose.position.latitude = projected_gps_point.lat;
+    vehicle_kinematics.geographic_pose.position.longitude = projected_gps_point.lon;
+    vehicle_kinematics.geographic_pose.position.altitude = projected_gps_point.ele;
+  } else if (map_projector_info_->type == "UTM") {
+    lanelet::GPSPoint position{
+      map_projector_info_->map_origin.latitude, map_projector_info_->map_origin.longitude};
+    lanelet::Origin origin{position};
+    lanelet::projection::UtmProjector projector{origin};
+    lanelet::GPSPoint projected_gps_point =
+      projector.reverse(toBasicPoint3dPt(kinematic_state_msgs_->pose.pose.position));
+    vehicle_kinematics.geographic_pose.header = kinematic_state_msgs_->header;
+    vehicle_kinematics.geographic_pose.header.frame_id = "global";
+    vehicle_kinematics.geographic_pose.position.latitude = projected_gps_point.lat;
+    vehicle_kinematics.geographic_pose.position.longitude = projected_gps_point.lon;
+    vehicle_kinematics.geographic_pose.position.altitude = projected_gps_point.ele;
+  }
```

Comment:
If the projection type is not supported, set header and set pose to NaN.

Reply:
fixed in [813b048](https://github.com/autowarefoundation/autoware.universe/pull/2930/commits/813b048edd79b610e773195a2ff7d3720f19f6a4)